cmd/routed-eni-cni-plugin/cni.go

```diff
@@ -407,41 +413,58 @@ func del(args *skel.CmdArgs, cniTypes typeswrapper.CNITYPES, grpcClient grpcwrap
-		return false, nil
+		log.Infof("no previous result found")
+		return false, false, nil
 	}
 
-	dummyIfaceName := networkutils.GeneratePodHostVethName(dummyVlanInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
+	dummyIfaceName := networkutils.GeneratePodHostVethName(dummyInterfacePrefix, string(k8sArgs.K8S_POD_NAMESPACE), string(k8sArgs.K8S_POD_NAME))
 	_, dummyIface, found := cniutils.FindInterfaceByName(prevResult.Interfaces, dummyIfaceName)
 	if !found {
-		return false, nil
+		log.Infof("interface %s not found in prevResult", dummyIfaceName)
+		return false, false, nil
 	}
+	// Pod VLAN ID is encoded in Interface.Mac
 	podVlanID, err := strconv.Atoi(dummyIface.Mac)
-	if err != nil || podVlanID == 0 {
-		return true, errors.Errorf("malformed vlanID in prevResult: %s", dummyIface.Mac)
-	}
-	if isNetnsEmpty(netNS) {
-		log.Infof("Ignoring TeardownPodENI as Netns is empty for SG pod:%s namespace: %s containerID:%s", k8sArgs.K8S_POD_NAME, k8sArgs.K8S_POD_NAMESPACE, k8sArgs.K8S_POD_INFRA_CONTAINER_ID)
-		return true, nil
+	// Note that a VLAN ID of 0 when conf.PodSGEnforcingMode is set does not necessarily mean an error, as the pod may have been created before the mode switch.
+	if err != nil {
+		return true, false, errors.Errorf("malformed vlanID in prevResult: %s", dummyIface.Mac)
 	}
+	branchEni := podVlanID != 0
 
 	containerIfaceIndex, _, found := cniutils.FindInterfaceByName(prevResult.Interfaces, contVethName)
 	if !found {
-		return false, errors.Errorf("cannot find contVethName %s in prevResult", contVethName)
+		return false, branchEni, errors.Errorf("cannot find contVethName %s in prevResult", contVethName)
 	}
 	containerIPs := cniutils.FindIPConfigsByIfaceIndex(prevResult.IPs, containerIfaceIndex)
 	if len(containerIPs) != 1 {
-		return false, errors.Errorf("found %d containerIP for %v in prevResult", len(containerIPs), contVethName)
+		return false, branchEni, errors.Errorf("found %d containerIP for %v in prevResult", len(containerIPs), contVethName)
 	}
 	containerIP := containerIPs[0].Address
 
-	if err := driverClient.TeardownBranchENIPodNetwork(&containerIP, podVlanID, conf.PodSGEnforcingMode, log); err != nil {
-		return true, err
+	if branchEni {
+		if isNetnsEmpty(netNS) {
+			log.Infof("Ignoring TeardownPodENI as Netns is empty for SG pod: %s namespace: %s containerID: %s", k8sArgs.K8S_POD_NAME, k8sArgs.K8S_POD_NAMESPACE, k8sArgs.K8S_POD_INFRA_CONTAINER_ID)
+			return true, branchEni, nil
+		}
+		err = driverClient.TeardownBranchENIPodNetwork(&containerIP, podVlanID, conf.PodSGEnforcingMode, log)
+	} else {
+		// deviceNumber is stored in interface Sandbox
+		var deviceNumber int
+		deviceNumber, err = strconv.Atoi(dummyIface.Sandbox)
+		if err != nil {
+			return false, branchEni, errors.Errorf("malformed deviceNumber in prevResult: %s", dummyIface.Sandbox)
+		}
+		err = driverClient.TeardownPodNetwork(&containerIP, deviceNumber, log)
+	}
+
+	if err != nil {
+		return true, branchEni, err
```

Comment:
since we return true here, does it mean we'll call ipamd to release IP but will skip teardownPodNetwork, and CNI won't retry due to we returned nil. seems we could leak some IP rules if TeardownPodNetwork failed for some reason.

```
if handled {
		return nil
	}
```

Reply:
The `TeardownPodNetwork` call that we make based on the IPAMD response will be the same as the call we made here, but it does seem safer to try again. I will change this to return `branchEni, branchEni, err`.